### PR TITLE
Refactor deploy

### PR DESCRIPTION
**Pull request by alan-barzilay**
This PR introduces a few refactorings to deploy.sh 

With the entry-point as the root of the file, I re-ordered the functions to follow the order upon which they are called in the entry-point, I believe this should make the file more readable and organized.
I also fixed a few minor issues such as few instances of file sourcing that weren't using the include function.

There were 3 functions for cleanup, they had a lot of overlap and were needlessly complicated so I merged them into one.

I also removed some unnecessary checks, the `is_kernel_root` check is done once in the entry-point function kernel_deploy so there is no need to remake this check in other functions later on

### CODE REVIEW

**Comment by alan-barzilay** [comment]
the shellcheck failures were already present and are being worked on by  @joaoseckler  and tracked in issue #323

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @alan-barzilay ,
This series looks really good. Can you include one extra commit that fixes the SC2124 issue in this file?

**Comment by alan-barzilay** [comment]
> Can you include one extra commit that fixes the SC2124 issue in this file?

Are you sure? @joaoseckler is working on resolving all cases of SC2124 in issue #318. Wouldn't a single commit fixing all occurrences be better? Also, if we end up needing to follow a certain specific pattern (or implement a helper function like in #319) it would be easier to make things consistent by doing a single commit

File: src/deploy.sh [hunk 1]
```diff
@@ -23,169 +23,336 @@ include "$KW_LIB_DIR/signal_manager.sh"
...
+    cleanup_after_deploy
+  fi
+}
 
-      # 3. Deploy: Execute script
-      local cmd="bash $REMOTE_KW_DEPLOY/deploy.sh --modules $release.tar"
-      cmd_remotely "$cmd" "$flag" "$remote" "$port"
-      ;;
-  esac
+function deploy_help()
```
Comment (rodrigosiqueira) [inline src/deploy.sh:145]: How about putting this `deploy_help` at the end of this file?

**Comment by rodrigosiqueira** [comment]
> > Can you include one extra commit that fixes the SC2124 issue in this file?
> 
> @rodrigosiqueira Are you sure? @joaoseckler is working on resolving all cases of SC2124 in issue #318. Wouldn't a single commit fixing all occurrences be better? Also, if we end up needing to follow a certain specific pattern (or implement a helper function like in #319) it would be easier to make things consistent by doing a single commit

Yes, I'm sure:
1. This series provides a set of basic refactors, and you actually fixed one of the shellcheck problems in this series.
2. If we have a single commit to fixing this issue, we can easily revert this change without affecting others. When we have a single commit that changes everything, it is terrible for revert because we will revert everywhere instead of a single point.

Honestly, I don't like these huge shellcheck fixes for the following reasons:
1. They are super hard to review.
2. We can easily miss something during the review and merge a bad code (see `is_kernel_root`).
3. It is terrible for bisect.
4. We can end up change multiple parts of the code, which messes with the git log.

Even though I don't like our current approach, I accepted because:
1. I don't want to make this shellcheck work endless.
2. We have five people paying attention to these types of changes.
3. I don't want to block other people's work since I have a limited amount of time with all of you (technically, after August, I'm on my own again).

**Comment by alan-barzilay** [comment]
> Yes, I'm sure

ok

File: src/deploy.sh (hunk 1 above)
Comment (alan-barzilay) [reply src/deploy.sh:145]: the whole point of reordering was to keep the functions in order of appearance, I would rather not move `deploy_help`

Comment (rodrigosiqueira) [reply src/deploy.sh:145]: What is the value of seeing such a function in the middle of the code? Technically, this function does not have any value from the development perspective, and if we keep it in the middle, it just looks like a "noise" when I'm looking for something more substantial. That was the exact situation that I have in other functions that have this helper function in the middle of the code.

Comment (alan-barzilay) [reply src/deploy.sh:145]: ok, I will move it

**Comment by alan-barzilay** [comment]
apparently pr #351 introduced a bug when naming the deploy test, the extra "s" in the name made it undiscoverable by run_tests.sh.

There are 2 tests failing, one of the cleanup function that I changed and needs to be adapted and one that I believe to me reminiscent from pr #351 as well. The kernel_install function cant be found because the plugins/kernel_install/utils.sh was never included. But including it makes everything  break because invoke shunit cant be found anymore.
I believe this is happening because now we have 2 file called utils.sh and this is breaking the code.

**Comment by rodrigosiqueira** [comment]
@alan-barzilay Could you be more specific about this bug? Because I just tested `kw d [-l|-u]`, and I just sent a PR that everything passes.

**Comment by alan-barzilay** [comment]
> @alan-barzilay Could you be more specific about this bug? Because I just tested kw d [-l|-u], and I just sent a PR that everything passes.

deploy is working, its tests are broken

**Comment by rodrigosiqueira** [comment]
Yeah... but what is exactly break? Because I just submitted this PR https://github.com/kworkflow/kworkflow/pull/358, and everything passes.

**Comment by alan-barzilay** [comment]
> Yeah... but what is exactly break? Because I just submitted this PR #358, and everything passes.

the test is not being run because of the file name

**Comment by rodrigosiqueira** [comment]
@alan-barzilay 
See https://github.com/kworkflow/kworkflow/pull/359  and https://app.travis-ci.com/github/kworkflow/kworkflow/builds/231853501#L1099 . I was running deploy_test locally, and everything was working as expected.

**Comment by joaoseckler** [comment]
Hey @alan-barzilay, @rodrigosiqueira! I just made a PR which will hopefully fix the issue with the inclusion of `utils.sh`, see #361. Maybe this could be rebased off that branch.

File: tests/deploy_test.sh [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+  cp -f "$MAINTAINERS_SAMPLE" "$FAKE_KERNEL"
+  cp -f "$CHECKPATH_EXT" "$FAKE_KERNEL/scripts/"
+
+  export preset_name="template_mkinitcpio.preset"
+  export test_path="$PWD/$FAKE_KERNEL"
+  export KW_PLUGINS_DIR="$test_path"
+  export KW_CACHE_DIR="$test_path"
+  export KW_ETC_DIR="$PWD/$SAMPLES_DIR/etc"
+  export DEPLOY_SCRIPT="$test_path/$kernel_install_path/deploy.sh"
+  export KW_PLUGINS_DIR="$PWD/src/plugins/"
```
Comment (magalilemes) [inline tests/deploy_test.sh:52]: Maybe have just one `KW_PLUGINS_DIR` export? (see line 48)

**Comment by alan-barzilay** [comment]
>  I was running deploy_test locally, and everything was working as expected.

@rodrigosiqueira  ok let me try to explain again in more detail because this involves at least 3 mistakes we made in the past, this test should really be working in unstable. It only broke here because I refactored deploy to use `include` instead of `. blabla.sh --sources-only.

There were 3 problems: first when implementing include we missed the imports that happened inside functions in `mk.sh`. Then we renamed utils to util.sh but we forgot there was already another utils.sh file and this is not something allowed, if  you try to import 2 files with the same name one of them won't be imported. Lastly when breaking down mk.sh we gave a bad name to the deploy tests, and it wasn't being recognized by the ./run_tests script anymore.

What happened is that I started using include inside the deploy.sh file, but only when I realized the tests weren't being run that this bug emerged. The install_kernel function  (not to be confused with the kernel_install function) needed for one of the tests couldn't be found because of the double import of utils.sh, this test file is the only occasion where both files are included at the same time.

I started fixing this issue in my last commit but #359 and now #361 seem to be addressing this.

TLDR: for this PR to pass the tests  #361 or something equivalent needs to be merged into unstable

File: tests/deploy_test.sh (hunk 2 above)
Comment (alan-barzilay) [reply tests/deploy_test.sh:52]: nice catch, I guess we can remove the first export

**Comment by rodrigosiqueira** [comment]
@alan-barzilay let me know when you have a new version.

**Comment by alan-barzilay** [comment]
> @alan-barzilay let me know when you have a new version.

@rodrigosiqueira this pr is ready for review

The only shellcheck error comes from a code that in #368  I argue that it should be disabled

**Comment by rodrigosiqueira** [comment]
@alan-barzilay could you fix this shellcheck issue?

**Comment by alan-barzilay** [comment]
> @alan-barzilay could you fix this shellcheck issue?

you fixed it by merging #386, I could rebase it into unstable to make it pass but I'm not at my computer right now, I would only be able to do it tomorrow night

**Comment by alan-barzilay** [comment]
@rodrigosiqueira everything ready

**Comment by rodrigosiqueira** [comment]
Merged, thanks.
